feat: initial Kubernetes dashboard

deboer-tim opened the pull request:
### What does this PR do?

Adds a dashboard to the Kubernetes section, as the first/default page. Follows the basic design from #9263 using current content/capability and without changing the nav structure.

The Kubernetes empty page is embedded in the dashboard, so now you'd see the top + empty page (but still no nav).

### Screenshot / video of UI

<img width="1022" alt="Screenshot 2024-10-24 at 1 27 53 PM" src="https://github.com/user-attachments/assets/878b1bd9-760f-4a16-be60-413f5d9c3b33">

### What issues does this PR fix or reference?

Fixes #9264.

### How to test this PR?

Go to Kubernetes section.

- [x] Tests are covering the bug fix or the new feature

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+let welcome = $derived($kubernetesCurrentContextState.error === NO_CURRENT_CONTEXT_ERROR);
+let currentContextName = $derived($kubernetesContexts.find(context => context.currentContext)?.name);
+
+let nodeCount = $state<number>(0);
+let deploymentCount = $state<number>(0);
+let serviceCount = $state<number>(0);
+let ingressCount = $state<number>(0);
+let routeCount = $state<number>(0);
+let pvcCount = $state<number>(0);
+let configMapCount = $state<number>(0);
+let secretCount = $state<number>(0);
```
nitpick: remove <number> typescript is smart enough to know it's a number

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+let unsubscribers: Unsubscriber[] = [];
+
+onMount(async () => {
+  unsubscribers.push(
+    kubernetesCurrentContextNodes.subscribe(nodes => {
+      nodeCount = nodes.length;
+    }),
+  );
```
suggestion: to make it simpler, replace it with $derived line 35
```
let nodeCount =$derived($nodes.length) 
```

benoitf replied:
and the same for the others it will reduce a lot of code

deboer-tim replied:
Was just rush-copying from existing code to start, I've fixed this now and it is much cleaner.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+  );
+});
+
+onDestroy(async () => {
+  unsubscribers.forEach(unsubscriber => unsubscriber());
+});
```
suggestion: using derived you'll be able to remove it

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+
+        <span>Get up and running by clicking one of the menu items!</span>
+
+        <Link class="place-self-start" on:click={() => window.openExternal('https://podman-desktop.io/docs/kubernetes')}>Kubernetes documentation</Link>
```
issue: use a function for the click (like openKubernetesDocumentation)

why ? b/c there is a linter rule saying you need to provide the returning type and it's not possible inlining

deboer-tim replied:
Fixed (in doc link, resource card, and guide cards.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+        <KubernetesDashboardResourceCard type='Services' Icon={ServiceIcon} count={serviceCount} link='/kubernetes/services'/>
+        <KubernetesDashboardResourceCard type='Ingresses & Routes' Icon={IngressRouteIcon} count={ingressCount + routeCount} link='/kubernetes/ingressesRoutes'/>
+        <KubernetesDashboardResourceCard type='Persistent Volume Claims' Icon={PvcIcon} count={pvcCount} link='/kubernetes/persistentvolumeclaims'/>
+        <KubernetesDashboardResourceCard type='Config Maps & Secrets' Icon={ConfigMapSecretIcon} count={configMapCount+secretCount} link='/kubernetes/configmapsSecrets'/>
```
issue: do not do maths here, just has a variable

deboer-tim replied:
Fixed via the $derived

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+        <KubernetesDashboardResourceCard type='Nodes' Icon={NodeIcon} count={nodeCount} link='/kubernetes/nodes'/>
+        <KubernetesDashboardResourceCard type='Deployments' Icon={DeploymentIcon} count={deploymentCount} link='/kubernetes/deployments'/>
+        <KubernetesDashboardResourceCard type='Services' Icon={ServiceIcon} count={serviceCount} link='/kubernetes/services'/>
+        <KubernetesDashboardResourceCard type='Ingresses & Routes' Icon={IngressRouteIcon} count={ingressCount + routeCount} link='/kubernetes/ingressesRoutes'/>
```
issue: do not do mathematics there, just use a variable

deboer-tim replied:
Ditto, fixed via the $derived

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+    {#if currentContextName}
+    <div class="text-xl">Metrics</div>
+    <div class="grid grid-cols-4 gap-4">
+        <KubernetesDashboardResourceCard type='Nodes' Icon={NodeIcon} count={nodeCount} link='/kubernetes/nodes'/>
```
suggestion (not-blocking)

it looks like we could just iterate over an array as only argument are changing

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.svelte:
```diff
@@ -0,0 +1,16 @@
+<script lang="ts">
+import { Button } from '@podman-desktop/ui-svelte';
+
+let {
+  title,
+  link,
+}: {
+  title: string;
+  link: string;
+} = $props();
```
todo: now we recommend to extract the properties to a separate Props interface

interface Props {
 title: string;
 link: string;
}

https://svelte.dev/docs/svelte/$props#Type-safety

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.svelte:
```diff
@@ -0,0 +1,16 @@
+<div class="flex flex-col gap-4 p-4 bg-[var(--pd-invert-content-card-bg)] rounded-md items-center">
+  <div class="font-bold text-[var(--pd-invert-content-card-text)]">{title}</div>
+  <Button onclick={() => window.openExternal(link)}>Read more</Button>
+</div>
```
issue: there is a missing EOF

deboer-tim replied:
Fixed manually... then I just noticed it was reverted again.

I'll fix this again before next commit, but biome isn't flagging or fixing this.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.svelte:
```diff
@@ -0,0 +1,16 @@
+
+<div class="flex flex-col gap-4 p-4 bg-[var(--pd-invert-content-card-bg)] rounded-md items-center">
+  <div class="font-bold text-[var(--pd-invert-content-card-text)]">{title}</div>
+  <Button onclick={() => window.openExternal(link)}>Read more</Button>
```
todo: use a function (I gave the reason upper)

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardResourceCard.svelte:
```diff
@@ -0,0 +1,29 @@
+import type { Component } from 'svelte';
+import { router } from 'tinro';
+
+let {
+  type,
+  Icon,
+  count,
+  link,
+}: {
+  type: string;
+  Icon: Component;
+  count: number;
+  link: string;
+} = $props();
```
todo: use an interface Props

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardResourceCard.svelte:
```diff
@@ -0,0 +1,29 @@
+} = $props();
+</script>
+
+<!-- svelte-ignore a11y_missing_attribute -->
+<!-- svelte-ignore a11y_no_static_element_interactions -->
+<!-- svelte-ignore a11y_click_events_have_key_events -->
```
issue: do not exclude a11y rules

deboer-tim replied:
Was copied from Link 😬. I don't remember if there was a reason we didn't use a button at the time, but I've switched this to a button and fixed (removed) these.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardResourceCard.svelte:
```diff
@@ -0,0 +1,29 @@
+        {count}
+        </div>
+    </div>
+  </a>
```
issue: missing EOF, linter/formatter is not working ?

deboer-tim replied:
Correct, as with above.

cdrage reviewed (commented):
![Screenshot 2024-10-24 at 8 19 06 AM](https://github.com/user-attachments/assets/d7c3d4e1-5d60-4f90-b519-c10886f29d83)

Shows all information from my cluster which is awesome!

Some suggestions:
- Hovering over the sections does not show the "click cursor"
- Should we have a "hover" border over the sections so we know it's clickable?
- 2 of my nodes are offline, but it shows 6 total, unsure if in UX/UI design we are showing... totals or are we showing what is working / not working?

cdrage commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,143 @@
+        <KubernetesDashboardResourceCard type='Services' Icon={ServiceIcon} count={serviceCount} link='/kubernetes/services'/>
+        <KubernetesDashboardResourceCard type='Ingresses & Routes' Icon={IngressRouteIcon} count={ingressCount + routeCount} link='/kubernetes/ingressesRoutes'/>
+        <KubernetesDashboardResourceCard type='Persistent Volume Claims' Icon={PvcIcon} count={pvcCount} link='/kubernetes/persistentvolumeclaims'/>
+        <KubernetesDashboardResourceCard type='Config Maps & Secrets' Icon={ConfigMapSecretIcon} count={configMapCount+secretCount} link='/kubernetes/configmapsSecrets'/>
```
```suggestion
        <KubernetesDashboardResourceCard type='ConfigMaps & Secrets' Icon={ConfigMapSecretIcon} count={configMapCount+secretCount} link='/kubernetes/configmapsSecrets'/>
```

ConfigMaps is one word

deboer-tim replied:
Fixed.

deboer-tim replied:
@cdrage PR has passed all tests, approved, and set to auto-merge but isn't. Seems to still be hung on this even though doesn't show at top right? This was fixed, please check again and review or merge. Thanks.

benoitf commented:
>Follows the basic design from https://github.com/containers/podman-desktop/issues/9263 using current content/capability and without changing the nav structure.

question what is the design to follow from https://github.com/containers/podman-desktop/issues/9263 ?

in the design articles = 3 columns, here it's two

question: in mockup, no Apply Yaml, here there is a button

for the cards
in mockup:
![image](https://github.com/user-attachments/assets/e6d97959-2dd7-4c30-b19d-fea6a5fcd9b2)

question: in the PR, icons are bigger, white, missing 'total' word
![image](https://github.com/user-attachments/assets/8fc5d41a-1cc6-4dde-9a89-4399cc12ccb4)

question: The order is not the same, also I'm wondering if there is an expected logical order or an alphabetical order

deboer-tim commented:
> > Follows the basic design from #9263 using current content/capability and without changing the nav structure.
> 
> question what is the design to follow from #9263 ?

I was following the latest mockup from Emma, in this comment:
https://github.com/containers/podman-desktop/issues/9263#issuecomment-2431903940

> in the design articles = 3 columns, here it's two

2 of the articles listed there are from RHD and wasn't sure we should include, so I found a second on podman-desktop.io blog. After the discussion today I'll revert to the three from the design.

> question: in mockup, no Apply Yaml, here there is a button

I was on the fence; had it here before the design came out because you can apply yaml that affects multiple resources and this is the only page you could actually see that.

> for the cards in mockup: 
> 
> question: in the PR, icons are bigger, white, text is bigger and missing 'total' word !

Yup, I still need to do those things, plus purple/obviously clickable.

> question: The order is not the same, also I'm wondering if there is an expected logical order or an alphabetical order

I am overriding design and saying the order should match the page order in the navigation. :) Maybe both need to change, but I think we had good reasons for (e.g.) putting Nodes before Deployments.

deboer-tim commented:
Rebased to fix merge conflict, add tests, fix many issues noted above, and improve UI.

deboer-tim commented:
> * Hovering over the sections does not show the "click cursor"

Fixed.

Resource cards: click cursor and hover effect everywhere, click anywhere.
Guide cards: hover effect everywhere, must click on the button. (copied what we do on the main dashboard)

> * Should we have a "hover" border over the sections so we know it's clickable?

I updated the icons to be purple as per the design. IMHO we use purple as a 'hover indicator' in enough places (plus the hover effect) that it is now more obvious it is clickable.

> * 2 of my nodes are offline, but it shows 6 total, unsure if in UX/UI design we are showing... totals or are we showing what is working / not working?

I added the word 'Total' to be closer to the design and clear that this is just the total number for now. I agree adding active count or donuts like the design would be nice, but I was planning to leave that for #9265.

deboer-tim commented:
I've responded to all the draft feedback, tests are basic but added and passing, and I wasn't planning to add anything else from the design for this sprint. Moving out of draft to kick off another round of feedback.

axel7083 commented on packages/renderer/src/lib/kube/KubernetesDashboard.spec.ts:
```diff
@@ -0,0 +1,117 @@
+
+// fake the window object
+beforeAll(() => {
+  //(window as any).events = eventsMock;
+  //(window as any).getConfigurationValue = vi.fn();
+  //(window as any).sendNavigationItems = vi.fn();
```
suggestion: to remove ?

deboer-tim replied:
Fixed.

axel7083 commented on packages/renderer/src/lib/kube/KubernetesDashboard.spec.ts:
```diff
@@ -0,0 +1,117 @@
+
+test('Verify basic page', async () => {
+  // mock no kubernetes resources
+  vi.mocked(kubeContextStore).kubernetesCurrentContextDeployments = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextServices = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextIngresses = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextRoutes = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextNodes = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextConfigMaps = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextSecrets = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextPersistentVolumeClaims = readable<KubernetesObject[]>([]);
+  vi.mocked(kubeContextStore).kubernetesCurrentContextState = readable<ContextGeneralState>({} as ContextGeneralState);
```
suggestion: those lines are the same than 83->91, they could be moved to a beforeEach

axel7083 commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,90 @@
+import KubernetesEmptyPage from './KubernetesEmptyPage.svelte';
+
+let noContexts = $derived($kubernetesCurrentContextState.error === NO_CURRENT_CONTEXT_ERROR);
+let currentContextName = $derived($kubernetesContexts.find(context => context.currentContext)?.name);
```
suggestion (non-blocking): I know typescript is able to infer currentContextName properly, but as a reader I might not know that it can be undefined

```suggestion
let currentContextName: string | undefined = $derived($kubernetesContexts.find(context => context.currentContext)?.name);
```

benoitf replied:
@axel7083 : if you read inside an IDE it's telling you the type 'string | undefined', so reading it outside of an IDE ?

axel7083 replied:
Reading in GitHub, yes the IDE is showing the type

axel7083 commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.spec.ts:
```diff
@@ -0,0 +1,63 @@
+
+// fake the window object
+beforeAll(() => {
+  (window as any).openExternal = openExternalMock;
```
suggestion (non-blocking): we could avoid having to defined `openExternalMock` which is not typed, by defining directly the mock on the function, then use `vi.mocked(window.openExternal).mockResolved...` or `expect(window.openExternal).toHaveBeenCalled()`

```suggestion
  (window.openExternal as unknown) = vi.fn();
```

axel7083 commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.svelte:
```diff
@@ -0,0 +1,19 @@
+
+let { title, link }: Props = $props();
+
+function openLink() {
+  window.openExternal(link);
+}
```
suggestion (non-blocking): svelte-check is suppose to complain that there is no return type, weird it says nothing

```suggestion
function openLink(): Promise<boolean> {
  return window.openExternal(link);
}
```

benoitf replied:
blocking for me (need return type + await or catch or return)

deboer-tim replied:
Fixed.

axel7083 commented on packages/renderer/src/lib/kube/KubernetesDashboardResourceCard.svelte:
```diff
@@ -0,0 +1,28 @@
+
+let { type, Icon, count, link }: Props = $props();
+
+function openLink() {
+  router.goto(link);
+}
```
```suggestion
function openLink(): void {
  router.goto(link);
}
```

deboer-tim replied:
Fixed.

axel7083 reviewed (approved):
Tested and LGTM! Great job :1st_place_medal: : this clearly enhance the Kubernetes section ! I put some non-blocking comments, but take my approval!

axel7083 commented:
suggestion (non-blocking, follow-up PR): maybe it would be interesting to have some telemetry on the resources card click, so we have a sense of how many users are using it to be redirected to the resources, something like

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.spec.ts:
```diff
@@ -0,0 +1,117 @@
+  //(window as any).events = eventsMock;
+  //(window as any).getConfigurationValue = vi.fn();
+  //(window as any).sendNavigationItems = vi.fn();
+  (window as any).openExternal = openExternalMock;
```
```suggestion
    Object.defineProperty(window, 'openExternal', {
    value: openExternalMock,
  });
```

suggestion: remove any usage

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,90 @@
+  $kubernetesCurrentContextConfigMaps.length + $kubernetesCurrentContextSecrets.length,
+);
+
+function openKubernetesDocumentation() {
+  window.openExternal('https://podman-desktop.io/docs/kubernetes');
+}
```
issue: missing return type

also openExternal is async so missing await or catch

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.spec.ts:
```diff
@@ -0,0 +1,63 @@
+
+// fake the window object
+beforeAll(() => {
+  (window as any).openExternal = openExternalMock;
```
AFAIK the best way to not use any or unknown is
```suggestion
  Object.defineProperty(window, 'openExternal', {
    value: openExternalMock,
  });
```

deboer-tim replied:
Fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardGuideCard.spec.ts:
```diff
@@ -0,0 +1,63 @@
+  expect(button).toBeInTheDocument();
+});
+
+test('Expect clicking words', async () => {
```
question: is it words or works ?

deboer-tim replied:
'Works', fixed.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboardResourceCard.spec.ts:
```diff
@@ -0,0 +1,56 @@
+  expect(count).toHaveClass('text-[var(--pd-invert-content-card-text)]');
+});
+
+test('Expect clicking words', async () => {
```
still the same question around words and works 🤔

deboer-tim replied:
Also 'works', also fixed.

benoitf reviewed (approved):
I've provided some minor feedback

still some overall questions
compared to the mockup
![image](https://github.com/user-attachments/assets/2df53339-a10b-4908-8f5a-bd8ed4b5d476)

icons are bigger there, and in the mockup digits are left aligned and here it's centered

I'm fine with any proposal, just wanted to confirm to either we keep the size, and move from left aligned to centered

probably a question to @ekidneyrh

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,90 @@
+
+                <div class="text-xl pt-2">Explore articles and blog posts</div>
+                <div class="grid grid-cols-3 gap-4">
+                    <KubernetesDashboardGuideCard title='Deploy and test Kubernetes containers using Podman Desktop' link='https://developers.redhat.com/articles/2023/06/09/deploy-and-test-kubernetes-containers-using-podman-desktop'/>
+                    <KubernetesDashboardGuideCard title='Working with Kubernetes in Podman Desktop' link='https://developers.redhat.com/articles/2023/11/06/working-kubernetes-podman-desktop'/>
+                    <KubernetesDashboardGuideCard title='Share your local podman images with the Kubernetes cluster' link='https://podman-desktop.io/blog/sharing-podman-images-with-kubernetes-cluster'/>
```
todo: create a follow up issue as it needs to be moved out to the backend, it should not be harcoded like that
it needs to be like for guided tour, in a json file

deboer-tim replied:
I was assuming this should be part of the existing follow up, #9265. I've added comments there to be explicit and note all of the current gaps I'm aware of.

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,90 @@
+                    </div>
+                {/if}
+
+                <div class="text-xl pt-2">Explore articles and blog posts</div>
```
question: there is no image while on mockup we had images ?

benoitf commented on packages/renderer/src/lib/kube/KubernetesDashboard.svelte:
```diff
@@ -0,0 +1,90 @@
+                    </div>
+                {/if}
+
+                <div class="text-xl pt-2">Explore articles and blog posts</div>
```
question: mockup = blogposts, here: blog posts

deboer-tim replied:
According to this I am just more parent than Tumblogger 😆:
https://english.stackexchange.com/questions/302723/blogpost-vs-blog-post#:~:text=It's%20a%20question%20of%20you,blog%20post'%20with%20your%20parents.

I tried to find other references - it seems like both are correct but blog post is still more common used on sites like Wikipedia.

deboer-tim commented:
I believe I've fixed all of the issues and responded to all the questions, and will merge since there are approvals. If I've missed anything urgent please feel free to ping or assign an issue to me, otherwise all remaining work to improve and match the design will be done as part of #9265.
